app/packages/viz3d/src/draco_loader.ts

```diff
@@ -0,0 +1,196 @@
+import * as three from "three"
+import * as draco from "draco3d"
+
+import * as draco_worker_lib from "./draco_worker"
+import draco_worker from "./draco_worker?worker"
+
+import * as worker_util from "./worker_util"
+export {LogLevel} from "./worker_util"
+
+//export { DracoTaskConfig } from "./draco_worker"
+
+// TODO: Improve performance.
+// This might be too slow for smooth UI.
+// As of Sun Feb 6 14:15:05 PST 2022 this can
+// load ~5mb point clouds (compressed to ~650kb) in ~65ms/each. 
+// ----
+// Fri Feb 11 12:55:44 PST 2022
+// Apple M1 Pro 8 core:
+// load ~5mb point clouds (compressed to ~650kb) in ~35ms/each
+// ----
+
+
+// TODO: Might be faster to do batch processing.
+// Measure overhead incurred from going back and 
+// forth from the worker to the client side.
+
+// TODO: Look into using some sort of shared ArrayBuffer 
+// which should speed this up significantly. 
+// (see: https://stackoverflow.com/questions/19152772/how-to-pass-large-data-to-web-workers)
+
+class DracoLoaderConfig {
+    public decodeConfig: draco_worker_lib.DracoTaskConfig;
+    public wasmBinary: ArrayBuffer;
+    public workerLimit: number = 4;
+    public logLevel: draco_worker_lib.LogLevel = draco_worker_lib.LogLevel.INFO;
+}
+
+export class DracoLoader extends worker_util.WorkerPool {
+    private _state: DracoLoaderConfig;
+
+    constructor ( state: DracoLoaderConfig) {
+       super(draco_worker, state.workerLimit);
+        this._state = state;
+        for (let i = 0; i < this._workers.length; i++){
+            this._workers[i].postMessage(draco_worker_lib.createInitRequest(i, this._state.wasmBinary, this._state.logLevel));
+        }
+    }
+
+    public onmessage(resolver: worker_util.Resolver, event: MessageEvent<any>) {
+        let message = event.data;
+        switch (message.type){
+            case draco_worker_lib.ResultMessageType.DECODE:
+                resolver.resolve(message.geometry);
+                break;
+            case draco_worker_lib.ResultMessageType.ERROR:
+                resolver.reject(message.message);
+                break;
+        }
+    }
+
+    public decodeGeometry (buffer: ArrayBuffer): Promise<any>{
+        return this.execute((worker, id) => {
+            // TODO: For some reason can't use transfer here. Attempts to do so
+            // throw "ArrayBuffer at index 0 is already detached" errors. Investigate this:
+            // worker.postMessage(draco_worker_lib.createDecodeRequest(id, buffer, this._state.decodeConfig), [buffer]);
+
+            worker.postMessage(draco_worker_lib.createDecodeRequest(id, buffer, this._state.decodeConfig));
```

Comment:
For some reason the transfer array isn't working here. Need to investigate further